components/server/src/api/workspace-service-api.ts

```diff
@@ -174,4 +182,60 @@ export class WorkspaceServiceAPI implements ServiceImpl<typeof WorkspaceServiceI
         response.workspace = this.apiConverter.toWorkspace(info);
         return response;
     }
+
+    async getWorkspaceDefaultImage(
+        req: GetWorkspaceDefaultImageRequest,
+        _: HandlerContext,
+    ): Promise<GetWorkspaceDefaultImageResponse> {
+        if (!req.workspaceId) {
+            throw new ApplicationError(ErrorCodes.BAD_REQUEST, "workspaceId is required");
+        }
+        const image = await this.workspaceService.getWorkspaceDefaultImage(ctxUserId(), req.workspaceId);
+        const response = new GetWorkspaceDefaultImageResponse({
+            defaultWorkspaceImage: image,
+        });
+        return response;
+    }
+
+    async sendHeartBeat(req: SendHeartBeatRequest, _: HandlerContext): Promise<SendHeartBeatResponse> {
+        if (!req.workspaceId) {
+            throw new ApplicationError(ErrorCodes.BAD_REQUEST, "workspaceId is required");
+        }
+        const info = await this.workspaceService.getWorkspace(ctxUserId(), req.workspaceId);
+        if (!info.latestInstance?.id || info.latestInstance.status.phase !== "running") {
+            throw new ApplicationError(ErrorCodes.BAD_REQUEST, "workspace is not running");
```

Comment:
BAD_REQUEST should be only used to validate arguments, it is FAILED_PRECONDITION